s2spy/_base_calendar.py

```diff
@@ -0,0 +1,222 @@
+        """
+        self._first_year = start
+        self._last_year = end
+        self._mapping = "years"
+        return self
+
+    def map_to_data(
+        self,
+        input_data: Union[pd.Series, pd.DataFrame, xr.Dataset, xr.DataArray],
+    ):
+        """Map the calendar to input data period.
+
+        Stores the first and last intervals of the input data to the calendar, so that
+        the intervals can cover the data to the greatest extent.
+
+        Args:
+            input_data: Input data for datetime mapping. Its index must be either
+                pandas.DatetimeIndex, or an xarray `time` coordinate with datetime
+                data.
+
+        Returns:
+            The calendar mapped to the input data period.
+        """
+        utils.check_input_data(input_data)
+
+        # check the datetime order of input data
+        if isinstance(input_data, PandasData):
+            self._first_timestamp = input_data.index.min()
+            self._last_timestamp = input_data.index.max()
+        else:
+            self._first_timestamp = pd.Timestamp(input_data.time.min().values)
+            self._last_timestamp = pd.Timestamp(input_data.time.max().values)
+
+        self._mapping = "data"
+
+        return self
+
+    def _year_range_from_timestamps(self):
+        map_first_year = self._first_timestamp.year
+        map_last_year = self._last_timestamp.year
+
+        # ensure that the input data could always cover the advent calendar
+        # last date check
+        if self._map_year(map_last_year).iloc[0].right > self._last_timestamp:
+            map_last_year -= 1
+        # first date check
+        if self._map_year(map_first_year).iloc[-1].left < self._first_timestamp:
+            map_first_year += 1
+
+        # map year(s) and generate year realized advent calendar
+        if map_last_year >= map_first_year:
+            self._first_year = map_first_year
+            self._last_year = map_last_year
+        else:
+            raise ValueError(
+                "The input data could not cover the target advent calendar."
+            )
+
+        return self
+
+    def _label_targets(self, intervals: pd.DataFrame) -> pd.DataFrame:
+        """Adds target labels to the header row of the intervals.
+
+        Args:
+            intervals (pd.Dataframe): Dataframe with intervals.
+
+        Returns:
+            pd.Dataframe: Dataframe with target periods labelled.
+        """
+        return intervals.rename(
+            columns={i: f"(target) {i}" for i in range(self.n_targets)}
+        )
+
+    def get_intervals(self) -> pd.DataFrame:
+        """Method to retrieve updated intervals from the Calendar object."""
+        if self._mapping is None:
+            raise ValueError(
+                "Cannot retrieve intervals without map_years or "
+                "map_to_data having configured the calendar."
+            )
+        if self._mapping == "data":
+            self._year_range_from_timestamps()
```

Comment:
```suggestion
            self._set_year_range_from_timestamps()
```
? Or: why not call this method directly from `map_to_data`?

Reply:
We could if we decide to disallow setting variables after initialization.

Currently the following code works without errors;
```py
>>> calendar = AdventCalendar(freq='30d', max_lag=15)
>>> calendar = calendar.map_to_data(input_data)
>>> calendar.show()
# User sees the calendar and decides on changing the max_lag, to avoid skipping years
>>> calendar.max_lag = 10
>>> calendar.show()
# Year ranges are re-generated from the min, max timestaps, thus the calendar will show correctly.
```
```

Reply:
Okay, let's follow-up in #69